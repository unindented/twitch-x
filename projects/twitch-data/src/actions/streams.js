import {getAction} from '../api'

export function loadTopStreams ({limit}) {
  const action = {
    type: 'LOAD_TOP_STREAMS'
  }
  return getAction(action, '/streams', {query: {limit}})
}
