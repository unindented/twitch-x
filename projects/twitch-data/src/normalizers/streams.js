import {normalize, schema} from 'normalizr'
import {valueOrNull} from './utils'

const stream = new schema.Entity('streams', {}, {
  idAttribute: '_id',
  processStrategy: (stream) => {
    return {
      id: valueOrNull(stream._id),
      status: valueOrNull(stream.channel.status),
      channel: valueOrNull(stream.channel.display_name),
      game: valueOrNull(stream.game),
      image: valueOrNull(stream.preview.template),
      viewers: valueOrNull(stream.viewers)
    }
  }
})

const root = [stream]

export function normalizeTopStreams (payload) {
  return normalize(payload.streams, root)
}

export {normalizeTopStreams as normalizeSearchStreams}
