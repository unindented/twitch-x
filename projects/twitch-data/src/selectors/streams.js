import {createSelector} from 'reselect'

export const getTopStreams = createSelector(
  (state) => state.data.streams.top,
  (state) => state.data.streams.byId,
  (topStreamIds, streamsById) => topStreamIds.map((id) => streamsById[id])
)

export const getTopStreamsForGame = createSelector(
  (state, props) => {
    const game = state.data.games.byId[props.id]
    return (game && game.topStreams) || []
  },
  (state) => state.data.streams.byId,
  (gameStreamIds, streamsById) => gameStreamIds.map((id) => streamsById[id])
)

export const getTopStreamsForCommunity = createSelector(
  (state, props) => {
    const community = state.data.communities.byId[props.id]
    return (community && community.topStreams) || []
  },
  (state) => state.data.streams.byId,
  (communityStreamIds, streamsById) => communityStreamIds.map((id) => streamsById[id])
)

export const getSearchStreams = createSelector(
  (state) => state.data.streams.search,
  (state) => state.data.streams.byId,
  (searchStreamIds, streamsById) => searchStreamIds.map((id) => streamsById[id])
)
