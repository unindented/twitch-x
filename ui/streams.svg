<?xml version="1.0" encoding="UTF-8"?>
<svg width="256px" height="256px" viewBox="0 0 256 256" version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
    <defs></defs>
    <g id="Assets" stroke="none" stroke-width="1" fill="none" fill-rule="evenodd">
        <g id="Streams" fill-rule="nonzero" fill="#FFFFFF">
            <path d="M181.333333,112 L181.333333,74.6666667 C181.333333,68.8 176.533333,64 170.666667,64 L42.6666667,64 C36.8,64 32,68.8 32,74.6666667 L32,181.333333 C32,187.2 36.8,192 42.6666667,192 L170.666667,192 C176.533333,192 181.333333,187.2 181.333333,181.333333 L181.333333,144 L224,186.666667 L224,69.3333333 L181.333333,112 Z" id="Shape"></path>
        </g>
    </g>
</svg>