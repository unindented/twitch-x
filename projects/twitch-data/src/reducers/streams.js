import update from 'immutability-helper'
import {normalizeTopStreams, normalizeSearchStreams} from '../normalizers/streams'

const initialState = {
  byId: {},
  top: [],
  search: []
}

export default (state = initialState, action = {}) => {
  switch (action.type) {
    case 'LOAD_TOP_STREAMS_SUCCESS': {
      const {entities, result} = normalizeTopStreams(action.payload.response)
      return update(state, {
        byId: {$merge: entities.streams},
        top: {$set: result}
      })
    }
    case 'LOAD_SEARCH_STREAMS_SUCCESS': {
      const {entities, result} = normalizeSearchStreams(action.payload.response)
      return update(state, {
        byId: {$merge: entities.streams},
        search: {$set: result}
      })
    }
  }

  return state
}
