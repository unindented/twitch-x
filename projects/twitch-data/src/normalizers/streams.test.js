import {normalizeTopStreams, normalizeSearchStreams} from './streams'

const streamsFixture = require('../__fixtures__/topStreams.json')
const searchStreamsFixture = require('../__fixtures__/searchStreams.json')

describe('normalizers/streams', () => {
  describe('.normalizeTopStreams', () => {
    it('normalizes the payload', () => {
      const result = normalizeTopStreams(streamsFixture)
      expect(result).toMatchSnapshot()
    })
  })

  describe('normalizers/streams', () => {
    describe('.normalizeSearchStreams', () => {
      it('normalizes the payload', () => {
        const result = normalizeSearchStreams(searchStreamsFixture)
        expect(result).toMatchSnapshot()
      })
    })
  })
})
