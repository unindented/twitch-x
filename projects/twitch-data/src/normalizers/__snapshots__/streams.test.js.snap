// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`normalizers/streams .normalizeTopStreams normalizes the payload 1`] = `
Object {
  "entities": Object {
    "streams": Object {
      "25774304576": Object {
        "channel": "PGL",
        "game": "Counter-Strike: Global Offensive",
        "id": 25774304576,
        "image": "https://static-cdn.jtvnw.net/previews-ttv/live_user_pgl-{width}x{height}.jpg",
        "status": "BIG vs Cloud9 (bo1) |  PGL Major Kraków Group Stage",
        "viewers": 277232,
      },
      "25774402160": Object {
        "channel": "Starladder5",
        "game": "Counter-Strike: Global Offensive",
        "id": 25774402160,
        "image": "https://static-cdn.jtvnw.net/previews-ttv/live_user_starladder5-{width}x{height}.jpg",
        "status": "LIVE: Vega Squadron vs Immortals | Round 2 | PGL Major Kraków Group Stage | by Sl4m & Strike",
        "viewers": 55614,
      },
      "25774417072": Object {
        "channel": "IzakOOO",
        "game": "Counter-Strike: Global Offensive",
        "id": 25774417072,
        "image": "https://static-cdn.jtvnw.net/previews-ttv/live_user_izakooo-{width}x{height}.jpg",
        "status": "[PGL Major] 18:00 BIG vs. Cloud9",
        "viewers": 31147,
      },
      "25774891744": Object {
        "channel": "imaqtpie",
        "game": "League of Legends",
        "id": 25774891744,
        "image": "https://static-cdn.jtvnw.net/previews-ttv/live_user_imaqtpie-{width}x{height}.jpg",
        "status": "wait a minute is this imaqtpies stream the one and only i gotta get here and watch him play league this guy does what no one else can hes so",
        "viewers": 19208,
      },
    },
  },
  "result": Array [
    25774304576,
    25774402160,
    25774417072,
    25774891744,
  ],
}
`;

exports[`normalizers/streams normalizers/streams .normalizeSearchStreams normalizes the payload 1`] = `
Object {
  "entities": Object {
    "streams": Object {
      "25768374736": Object {
        "channel": "ESL_SC2",
        "game": "StarCraft II",
        "id": 25768374736,
        "image": "https://static-cdn.jtvnw.net/previews-ttv/live_user_esl_sc2-{width}x{height}.jpg",
        "status": "RERUN: Welmu vs. Elazer (PvZ) - WCS Season 3 Challenger EU - Match 12",
        "viewers": 327,
      },
      "25771333072": Object {
        "channel": "OgamingSC2",
        "game": "StarCraft II",
        "id": 25771333072,
        "image": "https://static-cdn.jtvnw.net/previews-ttv/live_user_ogamingsc2-{width}x{height}.jpg",
        "status": "MarathOG - La suite du MarathOG",
        "viewers": 303,
      },
      "25774888880": Object {
        "channel": "wintergaming",
        "game": "StarCraft II",
        "id": 25774888880,
        "image": "https://static-cdn.jtvnw.net/previews-ttv/live_user_wintergaming-{width}x{height}.jpg",
        "status": "!Vodcast ❄ New sc2 players !startsc2 + !Guides for !terran/!zerg/!protoss  !youtube :)  Always commentary, never rage*",
        "viewers": 211,
      },
      "25775504816": Object {
        "channel": "LowkoTV",
        "game": "StarCraft II",
        "id": 25775504816,
        "image": "https://static-cdn.jtvnw.net/previews-ttv/live_user_lowkotv-{width}x{height}.jpg",
        "status": "Lowko :: StarCraft Dreams :: !prime",
        "viewers": 537,
      },
    },
  },
  "result": Array [
    25775504816,
    25768374736,
    25771333072,
    25774888880,
  ],
}
`;
